demo-orgs: Disable removing "invite required to join" until email set.

laurynmm opened the pull request:
Until a demo organization owner sets an email address, we want to restrict other users from joining the organization. Therefore, we disable changing the invite required setting for the organization until they set their email address. Otherwise, the owner could share the demo organization URL with someone and they could create an account via the homepage.

[Relevant CZO discussion](https://chat.zulip.org/#narrow/channel/3-backend/topic/demo.20organizations.3A.20setting.20org.20owner.20email/near/2169294)

**Notes**:
- I used the same tooltip handler that we have for the email visibility dropdown setting in the account and privacy tab in personal settings for demo organizations.
  - Alternatively, we could use the `tooltip_message` field for the settings checkbox in the template, and conditionally set that to either an empty string or the tooltip based on if the user's email is set (I believe I'd need to set the string in the `options` object in `web/src/admin.ts`).

**Screenshots and screen captures:**

| Before email is set | After email is set |
| --- | --- |
| ![Screenshot from 2025-05-13 15-29-02](https://github.com/user-attachments/assets/ad307076-ec29-494a-90a3-2172e282ef14) | ![Screenshot from 2025-05-13 15-30-13](https://github.com/user-attachments/assets/c8792486-eef6-440c-9007-d25de085eddd) |

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
The UI part looks fine to me.

laurynmm commented:
**Updates**:
- Added an initial commit that adds `check_demo_organization_state` as a function on the server side to check that no human owner account has an empty string as their email address. This commit adds that check to the already implemented code for converting a demo organization to a permanent organization via changing the realm's subdomain.
- Updated the commit that disables changing the `invite_required` realm setting to both update the web app organization admin UI and use the above check on the server-side code.
- No visual changes to the web app UI changes from the previous iteration. The implementation question about the tooltip is still valid.

laurynmm commented on zerver/views/realm.py:
```diff
@@ -85,6 +85,15 @@ def check_jitsi_url(value: str) -> str:
         raise JsonableError(_("{var_name} is not an allowed_type").format(var_name=var_name))
 
 
+def check_demo_organization_state(realm: Realm) -> None:
```
I wasn't sure about the best place to put this function. It could be added to the realm model. We could also add a `zerver/lib/demo_organizations.py` file or something like that.

timabbott replied:
`zerver/lib/demo_organizations.py` seems like a fine place.

timabbott replied:
Hmm, but I think I'd want a different name. Maybe `check_demo_organization_has_set_email`.

timabbott commented:
Looks good aside from the comments about naming that function.

laurynmm commented:
**Updates**:
- Added `zerver/lib/demo_organizations.py` and renamed the function for checking if the organization owner has set an email to `check_demo_organization_has_set_email` as suggested in the [comment thread above](https://github.com/zulip/zulip/pull/34621#discussion_r2097701751).

timabbott commented:
Merged, thanks @laurynmm!
